### PR TITLE
Add ability to log multiple key/value pairs at once.

### PR DESCRIPTION
**Pull request by elenadoty**
Bulk key/value logging will improve the performance of the SDK for developers who write out many custom keys and values at any given time by doing computationally expensive operations once for all pairs.

### CODE REVIEW

File: firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/common/CustomKeysAndValues.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+ * pairs with heterogenous value types.
+ */
+public class CustomKeysAndValues {
+
+  public final Map<String, String> keysAndValues;
+
+  static class Builder {
+
+    // Holds the converted pairs of custom keys and values.
+    Map<String, String> keysAndValues = new HashMap<String, String>();
```
Comment (rosariopfernandes) [inline firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/common/CustomKeysAndValues.java:32]: Won't this allow someone to reset it's value from another class?
(Eg. by calling `CustomKeysAndValues.Builder.keysAndValues = new HashMap<String, String>()`)

File: firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/common/CrashlyticsController.java [hunk 2]
```diff
@@ -437,6 +439,24 @@ void setCustomKey(String key, String value) {
...
+  void setCustomKeys(Map<String, String> keysAndValues) {
+    // Write all the key/value pairs before doing anything computationally expensive.
+    for (Map.Entry<String, String> entry : keysAndValues.entrySet()) {
+      try {
+        userMetadata.setCustomKey(entry.getKey(), entry.getValue());
+      } catch (IllegalArgumentException ex) {
+        if (context != null && CommonUtils.isAppDebuggable(context)) {
+          throw ex;
+        } else {
+          Logger.getLogger().e("Attempting to set custom attribute with null key, ignoring.");
```
Comment (blundell) [inline firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/common/CrashlyticsController.java:451]: you could log the key and or the value to help.

File: firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/common/CrashlyticsController.java [hunk 3]
```diff
@@ -361,7 +362,8 @@ boolean didCrashOnPreviousExecution() {
                                 if (appSettingsData == null) {
                                   Logger.getLogger()
                                       .w(
-                                          "Received null app settings at app startup. Cannot send cached reports");
+                                          "Received null app settings at app startup. Cannot send"
+                                              + " cached reports");
```
Comment (blundell) [inline firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/common/CrashlyticsController.java:366]: is this your local formatting rules?

File: firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/common/CrashlyticsController.java [hunk 4]
```diff
@@ -437,6 +439,24 @@ void setCustomKey(String key, String value) {
...
+    // Write all the key/value pairs before doing anything computationally expensive.
+    for (Map.Entry<String, String> entry : keysAndValues.entrySet()) {
+      try {
+        userMetadata.setCustomKey(entry.getKey(), entry.getValue());
+      } catch (IllegalArgumentException ex) {
+        if (context != null && CommonUtils.isAppDebuggable(context)) {
+          throw ex;
+        } else {
+          Logger.getLogger().e("Attempting to set custom attribute with null key, ignoring.");
+          return;
```
Comment (blundell) [inline firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/common/CrashlyticsController.java:452]: would you want to `continue` to write the rest? this cuts out the for loop at the problem

File: firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/common/CrashlyticsCore.java [hunk 5]
```diff
@@ -329,6 +330,21 @@ public void setCustomKey(String key, String value) {
...
+  /**
+   * Sets multiple values to be associated with given keys for your crash data. This method should
+   * be used instead of setCustomKey when many different key/value pairs are to be set at the same
+   * time in order to optimize the process of writing out the data. The key/value pairs will be
+   * reported with any crash that occurs in this session. A maximum of 64 key/value pairs can be
+   * stored for any type. New keys added over that limit will be ignored. Keys and values are
+   * trimmed ({@link String#trim()}), and keys or values that exceed 1024 characters will be
+   * truncated.
+   *
+   * @throws NullPointerException if key is null.
```
Comment (blundell) [inline firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/common/CrashlyticsCore.java:342]: do you mean `if keysAndValues is null` ?

File: firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/common/CustomKeysAndValues.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+import androidx.annotation.NonNull;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Helper class which handles the storage and conversion to strings of key/value pairs with
+ * heterogenous value types.
+ */
+public class CustomKeysAndValues {
```
Comment (blundell) [inline firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/common/CustomKeysAndValues.java:25]: this seems to be in the main sourceset but only used in the androidTest source set. What's the reason?

**Pull request closed without merging**